syntax = "proto3";

package statsservice;

message GetPostStatsRequest {
  int64 post_id = 1;
}

message GetPostStatsResponse {
  int64 post_id = 1;
  int64 like_count = 2;
  int64 view_count = 3;
  int64 comment_count = 4;
}

message GetPostViewDynamicsRequest {
  int64 post_id = 1;
}

message PostViewDay {
  string date = 1;
  int64 view_count = 2;
}

message GetPostViewDynamicsResponse {
  repeated PostViewDay days = 1;
}

message GetPostLikeDynamicsRequest {
  int64 post_id = 1;
}

message PostLikeDay {
  string date = 1;
  int64 like_count = 2;
}

message GetPostLikeDynamicsResponse {
  repeated PostLikeDay days = 1;
}

message GetPostCommentDynamicsRequest {
  int64 post_id = 1;
}

message PostCommentDay {
  string date = 1;
  int64 comment_count = 2;
}

message GetPostCommentDynamicsResponse {
  repeated PostCommentDay days = 1;
}

enum TopMetric {
  TOP_LIKES = 0;
  TOP_COMMENTS = 1;
  TOP_VIEWS = 2;
}

message GetTopPostsRequest {
  TopMetric metric = 1;
}

message TopPost {
  int64 post_id = 1;
  int64 count = 2;
}

message GetTopPostsResponse {
  repeated TopPost posts = 1;
}

message GetTopUsersRequest {
  TopMetric metric = 1;
}

message TopUser {
  int64 user_id = 1;
  int64 count = 2;
}

message GetTopUsersResponse {
  repeated TopUser users = 1;
}

service StatsService {
  rpc GetPostStats(GetPostStatsRequest) returns (GetPostStatsResponse);
  rpc GetPostViewDynamics(GetPostViewDynamicsRequest) returns (GetPostViewDynamicsResponse);
  rpc GetPostLikeDynamics(GetPostLikeDynamicsRequest) returns (GetPostLikeDynamicsResponse);
  rpc GetPostCommentDynamics(GetPostCommentDynamicsRequest) returns (GetPostCommentDynamicsResponse);
  rpc GetTopPosts(GetTopPostsRequest) returns (GetTopPostsResponse);
  rpc GetTopUsers(GetTopUsersRequest) returns (GetTopUsersResponse);
} 